Closes #1139: Clustered cache manager configuration documentation

albinsuresh opened the pull request:
An image file is used in this PR. I'll try to replace that with an asciiart if possible.

GaryWKeim commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -47,14 +84,31 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
-After starting the Terracotta server (Refer section: <<Starting Terracotta server>>), Cache manager with clustering capabilities can be configured with server side resource(s) and in turn can be used by cache(s) :
+After starting the Terracotta server (Refer section: <<Starting Terracotta server>>), Cache manager with clustering capabilities can be configured with server side resource(s) and in turn can be used by cache(s).
+The following sample; unlike the previous one; uses a broader clustering service configuration that lets you configure more things at the cache manager level.
+Before we jump it configuration samples, we'll define certain terms that are used in the reminder of this document.
```
jump *to* configuration samples
*remainder* of this document

GaryWKeim commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -65,12 +119,11 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
-<4> Adds a resource pool for the cache manager with the specified name(*resource-pool-a*) and size(*128B*) consumed out of a named server storage resource(*secondary-server-resource*)
-    already existing on the server.
+    The server storage resource identifiers will be used to create resource pools.
+<4> Adds a resource pool for the cache manager with the specified name(*resource-pool-a*) and size(*128B*) consumed out of the  named server resource *secondary-server-resource*
```
Can we change the code to use some more realistically sized pools?

albinsuresh commented:
Reverted the ascii art to image file as the rendering of the ascii art was not looking good. Additional gradle plugins may have to be configured to get that rendering right. Will switch back to ascii art once the build part gets resolved.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,45 @@ ifdef::notBuildingForSite[]
+----
+
+The above configuration defines two named server resources: *primary-server-resource* and *secondary-server-resource*.
+The clients(cache managers) can consume storage from these named resources by configuring the cache manager accordingly.
```
Missing a space before parenthesis

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -37,7 +58,23 @@ Check for the below `INFO` log to confirm if the server started successfully,
 == Creating cache manager with clustering capabilities
 
 After starting the Terracotta server (Refer section: <<Starting Terracotta server>>), you can now proceed to create the cache manager.
```
Could you change that to a link to the section as a hyper link around "starting the Terracotta server"? Also when doing such a link, please add an explicit session id.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -47,14 +84,36 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
 == Cache manager configuration and usage of server side resources
 
-After starting the Terracotta server (Refer section: <<Starting Terracotta server>>), Cache manager with clustering capabilities can be configured with server side resource(s) and in turn can be used by cache(s) :
+After starting the Terracotta server (Refer section: <<Starting Terracotta server>>), Cache manager with clustering capabilities can be configured with server side resource(s) and in turn can be used by cache(s).
```
Same remark as above for the link

ljacomet replied:
Incorrect capital C in Cache manager

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -47,14 +84,36 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
+
+Server resource::
+Server resources are storage resources defined at the server.
+*primary-server-resource* and *secondary-server-resource* are two named server resources defined in the server configuration from previous section.
```
Replace *bold* usage by `code`

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -47,14 +84,36 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
+
+Here is a pictorial representation of server resources and resource pools:
+
+image::storagePools.png[]
```
This does not render in a browser when doing `./gradlew asciidoctor`. Any chance it can be fixed?

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -65,18 +124,17 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
-    Since the server storage identifier is not explicitly passed, this resource pool will be consumed out of default server resource provided in Step 3.
-    This demonstrates that a cache manager with clustering support can have multiple resource pools created out of several server storage resources.
+    The server storage resource identifiers will be used to create resource pools.
+<4> Adds a resource pool for the cache manager with the specified name(*resource-pool-a*) and size(*28MB*) consumed out of the  named server resource *secondary-server-resource*
```
replace **bold** with `code` and space before parenthesis.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -65,18 +124,17 @@ include::{sourcedir}/clustered/client/src/test/java/org/ehcache/clustered/client
-    This demonstrates that a cache manager with clustering support can have multiple resource pools created out of several server storage resources.
+    The server storage resource identifiers will be used to create resource pools.
+<4> Adds a resource pool for the cache manager with the specified name(*resource-pool-a*) and size(*28MB*) consumed out of the  named server resource *secondary-server-resource*
+<5> Adds a resource pool for the cache manager with the specified name(*resource-pool-b*) and size(*32MB*).
```
These descriptions are too long for callouts. There should be more upfront definition and explanation and the callouts only serve as illustration.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,46 @@ ifdef::notBuildingForSite[]
+include::{sourcedir}/clustered/client/src/test/resources/configs/docs/tc-config.xml[]
+----
+
+The above configuration defines two named _server off-heap resources_: *primary-server-resource* and *secondary-server-resource*.
```
This should use callouts

ljacomet replied:
And again, the italic is ok, the bold is wrong - you are referencing something from a source or config - use backticks for that

albinsuresh commented:
Addressed the review feedback on the usage of bold characters instead of back ticks.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -36,50 +62,92 @@ Check for the below `INFO` log to confirm if the server started successfully,
+The URI can be split into 3 logical units.
+We will analyze what each unit in the above URI represents.
+
+. *terracotta://localhost:9510*: the URI where the Terracotta server instance is running
```
Again, code extract - backtick instead of bold

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -36,50 +62,92 @@ Check for the below `INFO` log to confirm if the server started successfully,
+`primary-server-resource` and `secondary-server-resource` are two named server off-heap resources defined in the server configuration from previous section.
+Clustered caches can have fixed amount of storage allocated directly from these resources.
+
+Resource pool::
```
You are missing a similar section on cache fixed pool since the picture below contains some.

albinsuresh replied:
I was hoping that line 106 would cover that.

GaryWKeim commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -36,50 +62,92 @@ Check for the below `INFO` log to confirm if the server started successfully,
+That is, if two caches are using a shared resource pool, the data of each cache is still isolated.
+One can not access the data of the other.
+They are just sharing the same storage resource.
+What it means is that, when shared caches hit the capacity of their shared resource pool, and when one tries to expand, it will expand by shrinking others' storage resource.
```
I think this sentence can be made clearer, but I also wonder if it's correct as is. If a shared pool is at capacity, then a putAll is executed on cache Foo, is it absolutely true that only the other caches pool usage will be lowered (mappings evicted) or could mappings in Foo also be removed to accommodate the new mappings.

albinsuresh commented:
Refactored the whole document to extract out the clustering concepts from configuration samples. I'd strongly recommend re-reading the whole thing as I've modified, added and deleted content from the previous iterations.

albinsuresh commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,77 +9,132 @@ ifdef::notBuildingForSite[]
+
+In this section we discuss some Terracotta clustering terms and concepts that users need to understand before creating cache managers with clustering support.
+
+Clustered storage::
```
During the sprint demos I heard @cljohnso referring to this as `clustered storage manager`. That sounds better to me. Will change it if no one disagrees to that.

ljacomet replied:
I am wondering if we should not simply use _Ehcache entity_. Problem with _clustered storage manager_ is that it does not limit itself to Ehcache just by name.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,77 +9,132 @@ ifdef::notBuildingForSite[]
 To enable Clustering with Terracotta, firstly you will have to start the http://terracotta.org[Terracotta server] configured with clustered storage.
 And then need a cache manager with clustering capabilities so that the caches can consume the clustered storage.
 
+Detailed instructions on how to configure and start a Terracotta server array can be found elsewhere.
```
I would move this to the next section as it relates to TC server config and starting.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,77 +9,132 @@ ifdef::notBuildingForSite[]
 Check for the below `INFO` log to confirm if the server started successfully,
 `Terracotta Server instance has started up as ACTIVE node on 0:0:0:0:0:0:0:0:9510 successfully, and is now ready for work.`
 
+== Clustering concepts
```
I would move this section below the introduction.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,77 +9,132 @@ ifdef::notBuildingForSite[]
+
+Server off-heap resource::
+Server off-heap resources are storage resources defined at the server.
+`primary-server-resource` and `secondary-server-resource` are two named server off-heap resources defined in the server configuration from previous section.
```
once this moves before the starting TC server example, the reference should be maintained the other way - from the start section, indicate that the resources are _server off-heap resources_.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,77 +9,132 @@ ifdef::notBuildingForSite[]
+A clustered storage instance at the server side is identified by a unique _clustered storage identifier_.
+Multiple clients can identify and connect to the same clustered storage instance using this unique identifier.
+
+Server off-heap resource::
```
I would move these first, as they are server concept and not something _internal_ to the entity.

ljacomet commented:
Made another batch of comments but really like where this is going.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+Multiple clients can identify and connect to the same clustered entity using this unique identifier.
+
+Fixed pool::
+Fixed pools are fixed-amount storage pools allocated to clients.
```
Nothing is allocated to clients - it is allocated to clustered tier of a cache.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+Server off-heap resources are storage resources defined at the server.
+`primary-server-resource` and `secondary-server-resource` are two named server off-heap resources defined in the sample
+server configuration in the next section on <<start-server, starting the Terracotta server>>.
+Clients can get their storage resources allocated from such instances.
```
There is no resource allocated to clients - allocation is either for the entity - shared pools - or for the cache clustered tier.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+Fixed pool::
+Fixed pools are fixed-amount storage pools allocated to clients.
+A fixed amount of storage is allocated directly from server off-heap resources to these pools.
+Clients will have exclusive access to the storage resources in the fixed pool that they're configured with.
```
Again, client is not the right term.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+Clients will have exclusive access to the storage resources in the fixed pool that they're configured with.
+
+Shared pool::
+Shared pools are again fixed-amount storage pools, but can be shared by multiple clients.
```
shared by multiple cache clustered tiers.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+Shared pools are again fixed-amount storage pools, but can be shared by multiple clients.
+As in the case of fixed pools, shared pools are also carved out from server off-heap resources.
+The storage available in these shared pools are strictly shared.
+That is, no client can ask for a fixed amount of storage from these shared pools.
```
again client -> cache clustered tier

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+As in the case of fixed pools, shared pools are also carved out from server off-heap resources.
+The storage available in these shared pools are strictly shared.
+That is, no client can ask for a fixed amount of storage from these shared pools.
+If a client needs a fixed amount of storage that it is not willing to share, then it must get a fixed pool allocated.
```
client

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+If a client needs a fixed amount of storage that it is not willing to share, then it must get a fixed pool allocated.
+
+Sharing of storage using shared pools here doesn't mean that the data is shared.
+That is, if two clients are using a shared resource pool, the data of each client is still isolated but the underlying storage is shared.
```
client again

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+
+Sharing of storage using shared pools here doesn't mean that the data is shared.
+That is, if two clients are using a shared resource pool, the data of each client is still isolated but the underlying storage is shared.
+What it means is that, when shared clients hit the capacity of their shared resource pool, and when one tries to expand, it could expand by shrinking others' storage resource by evicting its entries if required.
```
and here too

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,84 @@ ifdef::notBuildingForSite[]
+
+Server off-heap resource::
+Server off-heap resources are storage resources defined at the server.
+`primary-server-resource` and `secondary-server-resource` are two named server off-heap resources defined in the sample
```
Not too convinced with this forward reference.

albinsuresh commented:
Updated the PR by addressing the review comments: majorly to avoid using the term "clients" to refer to clustered cache managers and caches. The last two commits or even future commits will all be squashed to one once the content is deemed good to go.

ljacomet commented on docs/src/docs/asciidoc/user/clustered-cache.adoc:
```diff
@@ -9,24 +9,83 @@ ifdef::notBuildingForSite[]
+
+Server off-heap resource::
+Server off-heap resources are storage resources defined at the server.
+Caches can get the storage resources for their clustered tiers allocated from these server off-heap resources.
```
I would drop the first occurence of _resources_ from this sentence.
